server/src/main/java/umm3601/Server.java

```diff
@@ -26,39 +24,47 @@ public static void main(String[] args) {
     String databaseName = System.getenv().getOrDefault("MONGO_DB", "dev");
 
     // Setup the MongoDB client object with the information we set earlier
-    MongoClient mongoClient = MongoClients.create(
-      MongoClientSettings.builder()
-      .applyToClusterSettings(builder ->
-        builder.hosts(Arrays.asList(new ServerAddress(mongoAddr))))
-      .build());
+    MongoClient mongoClient
+      = MongoClients.create(MongoClientSettings
+        .builder()
+        .applyToClusterSettings(builder -> builder.hosts(Arrays.asList(new ServerAddress(mongoAddr))))
+        .build());
 
     // Get the database
-    database = mongoClient.getDatabase(databaseName);
+    MongoDatabase database = mongoClient.getDatabase(databaseName);
 
     // Initialize dependencies
     UserController userController = new UserController(database);
-    //UserRequestHandler userRequestHandler = new UserRequestHandler(userController);
-
-    Javalin server = Javalin.create().start(4567);
 
-    // Simple example route
-    server.get("hello", ctx -> ctx.result("Hello World"));
+    Javalin server = Javalin.create();
+    /*
```

Comment:
There were warnings about the fact that our `MongoClient` isn't necessarily closed if something Bad happens and, for example, the server crashes.

The recommendations were to use the try-with-resources or a `try` with a `finally` block, but those don't work here. The `MongoClient` variable has to be "live" the entire time that the server is up, so we can't shut it down anywhere in `main()`.

This adds server event handlers that close the client when either the server fails to actually start, or the server is stopping. I don't know if those are the right ones or enough, so feedback would certainly be appreciated.

I also don't know if this is several more lines of code and additional complication that doesn't actually make anything better, so thoughts there would be good as well.

Reply:
Oh, I clearly put this on the wrong line. The relevant code is:

```java
Javalin server = Javalin.create();
    /*
     * We want to shut the `mongoClient` down if the server either
     * fails to start, or when it's shutting down for whatever reason.
     * Since the mongClient needs to be available throughout the
     * life of the server, the only way to do this is to wait for
     * these events and close it then.
     */
    server.events(event -> {
      event.serverStartFailed(mongoClient::close);
      event.serverStopping(mongoClient::close);
    });
    server.start(4567);
```

Reply:
Looks reasonable!

It seems like like `MongoClient.close()` does some nice things for us. Here's its JavaDoc:

```java
    /**
     * Close the client, which will close all underlying cached resources, including, for example,
     * sockets and background monitoring threads.
     */
    void close();
```

Reply:
Is there a reason you choose `serverStopping` rather than `serverStopped`?

Right now `server.stop()` is never called so that event hook never happens anyway. We can add this to actually shut down cleanly:
```java
Runtime.getRuntime().addShutdownHook(new Thread(() -> {
	app.stop();
}));
```